### PR TITLE
Add describe datamodel command

### PR DESCRIPTION
**Pull request by gregertw**
This pull request adds a new command called `describe datamodel` to the `cdf-tk` tool. 

This command allows users to describe the characteristics of a data model given the space name and data model name. It retrieves information about the space, containers, data model, views, and relations. This command enhances the functionality of the `cdf-tk` tool by providing a convenient way to explore and document data models in CDF.

### CODE REVIEW

File: cognite_toolkit/cdf.py [hunk 1]
```diff
@@ -52,7 +53,11 @@
 auth_app = typer.Typer(
     pretty_exceptions_short=False, pretty_exceptions_show_locals=False, pretty_exceptions_enable=False
 )
+describe_app = typer.Typer(
```
Comment (doctrino) [inline cognite_toolkit/cdf.py:56]: What is the motivation for introducing another app?

File: tests/test_approval_modules_snapshots/cdf_oid_example_data.yaml [hunk 2]
```diff
@@ -27,7 +27,7 @@ Datapoints:
   - pi_163582
   - pi_160781
   - pi_191092
-  dataframe: 29026408060742985162782212323348235008860481513847239798105871600423802129032
+  dataframe: 14540507320479421898910558370683995785631409828267798058294586154301552423325
```
Comment (doctrino) [inline tests/test_approval_modules_snapshots/cdf_oid_example_data.yaml:30]: Any reason why this is changing?

File: cognite_toolkit/cdf_tk/describe.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        data_model = client.data_modeling.data_models.list(
+            space=space_name, include_global=True, inline_views=True, limit=None
+        )
+    except Exception as e:
+        print(f"Failed to retrieve data model {model_name} in space {space_name}.")
+        print(e)
+        return
+    if len(data_model) == 0:
+        print(f"Failed to retrieve data model {model_name} in space {space_name}.")
+        return
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:67]: Future `mypy` will complain
```suggestion
        return None
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+            table.add_row("Name", str(space.name))
+            table.add_row("Description", str(space.description))
+            table.add_row("Created time", str(datetime.datetime.fromtimestamp(space.created_time / 1000)))
+            table.add_row("Last updated time", str(datetime.datetime.fromtimestamp(space.last_updated_time / 1000)))
+    try:
+        containers = client.data_modeling.containers.list(space=space_name, include_global=True, limit=None)
+    except Exception as e:
+        print(f"Failed to retrieve containers for data model {model_name}.")
+        print(e)
+        return
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:48]:
```suggestion
        return None
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    container_list = [(space_name, c.external_id) for c in containers.data]
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:49]: We have a shorthand for this `containres.as_ids()` However, I do not understand why you want to do that, as you can just iterate through the containers. See next comment

File: cognite_toolkit/cdf_tk/describe.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    containers_str = []
+    for c in container_list:
+        containers_str.append(f"{c[1]}\n")
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:52]:
```suggestion
    containers_str = "\n".join([c.external_id for c in containers])
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    table.add_row(f"Containers ({len(containers_str)})", "".join(containers_str))
+    print(table)
+    try:
+        data_model = client.data_modeling.data_models.list(
+            space=space_name, include_global=True, inline_views=True, limit=None
+        )
+    except Exception as e:
+        print(f"Failed to retrieve data model {model_name} in space {space_name}.")
+        print(e)
+        return
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:64]:
```suggestion
        return None
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    if len(data_model) == 0:
+        print(f"Failed to retrieve data model {model_name} in space {space_name}.")
+        return
+    model_name = data_model.data[0].name
+    if len(data_model.data) > 1:
+        print(f"Found {len(data_model.data)} data models in space {space_name}.")
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:70]: No need to access the data attribute here, `CogniteResourceList` are lists. 
```suggestion
        print(f"Found {len(data_model)} data models in space {space_name}.")
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+        )
+    except Exception as e:
+        print(f"Failed to retrieve data model {model_name} in space {space_name}.")
+        print(e)
+        return
+    if len(data_model) == 0:
+        print(f"Failed to retrieve data model {model_name} in space {space_name}.")
+        return
+    model_name = data_model.data[0].name
+    if len(data_model.data) > 1:
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:69]:
```suggestion
    if len(data_model) > 1:
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+        print(f"  Only describing the first one ({model_name}).")
+        print("  Use the --data-model flag to specify which data model to describe.")
+
+    table = Table(title=f"Data model {model_name} in space {space_name}")
+    table.add_column("Info", justify="right")
+    table.add_column("Value", justify="left", style="green")
+    table.add_row("Description", str(data_model.data[0].description))
+    table.add_row(
+        "Version",
+        str(data_model.data[0].version),
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:80]:
```suggestion
        str(data_model[0].version),
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+    )
+    table.add_row(
+        "Global",
+        "True" if data_model.data[0].is_global else "False",
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:84]:
```suggestion
        "True" if data_model[0].is_global else "False",
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+    )
+    table.add_row("Created time", str(datetime.datetime.fromtimestamp(data_model.data[0].created_time / 1000)))
+    table.add_row(
+        "Last updated time", str(datetime.datetime.fromtimestamp(data_model.data[0].last_updated_time / 1000))
+    )
+    views = data_model.data[0].views
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:90]:
```suggestion
    views = data_model[0].views
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+    table.add_row("List of views", "".join(view_names))
+    print(table)
+    for view in views:
+        table = Table(title=f"View {view.external_id}, version {view.version} in space {space_name}")
+        table.add_column("Info", justify="left")
+        table.add_column("Value", justify="left", style="green")
+        table.add_row("Number of properties", str(len(view.properties)))
+        table.add_row("Used for", str(view.used_for))
+
+        implements_str = [f"{i}\n" for i in view.implements or []]
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:104]: Should we do the same as for views and containers?
```suggestion
        implements_str = [f"{i.external_id}\n" for i in view.implements or []]
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+        if len(implements_str) > 0:
+            implements_str[-1] = implements_str[-1][0:-1]
+        table.add_row("Implements", "".join(implements_str))
+        properties = [f"{p}\n" for p in view.properties.keys()]
+        if len(properties) > 0:
+            properties[-1] = properties[-1][0:-1]
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:110]:
```suggestion
        properties = "\n".join(view.properties.keys())
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+    )
+    table.add_row("Created time", str(datetime.datetime.fromtimestamp(data_model.data[0].created_time / 1000)))
+    table.add_row(
+        "Last updated time", str(datetime.datetime.fromtimestamp(data_model.data[0].last_updated_time / 1000))
+    )
+    views = data_model.data[0].views
+    table.add_row("Number of views", str(len(views)))
+    view_names = [f"{v.external_id}\n" for v in views]
+    if len(view_names) > 0:
+        view_names[-1] = view_names[-1][0:-1]
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:94]:
```suggestion
    view_names = "\n".join([v.external_id for v in views])
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+    except Exception as e:
+        print(f"Failed to retrieve containers for data model {model_name}.")
+        print(e)
+        return
+    container_list = [(space_name, c.external_id) for c in containers.data]
+    containers_str = []
+    for c in container_list:
+        containers_str.append(f"{c[1]}\n")
+    if len(containers_str) > 0:
+        containers_str[-1] = containers_str[-1][0:-1]
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:54]:
```suggestion
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+                )
+            elif type(edge_type.type) is DirectRelationReference:
+                nr_of_edge_relations += 1
+                edge_relations_str.append(
+                    f"{p} -- {edge_type.direction} --> ({edge_type.source.space}, {edge_type.source.external_id}, {edge_type.source.version})\n"
+                )
+        if len(direct_relations_str) > 0:
+            direct_relations_str[-1] = direct_relations_str[-1][0:-1]
+        if len(edge_relations_str) > 0:
+            edge_relations_str[-1] = edge_relations_str[-1][0:-1]
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:133]:
```suggestion
        nr_of_direct_relations = len(direct_relations_str)
        nr_of_edge_relations = len(edge_relations_str)
        direct_relations_str = "\n".join(direct_relations_str)
        edge_relations_str = "\n".join(edge_relations_str)
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+                if edge_type.source is None:
+                    direct_relations_str.append(f"{p} --> no source")
+                    continue
+                direct_relations_str.append(
+                    f"{p} --> ({edge_type.source.space}, {edge_type.source.external_id}, {edge_type.source.version})\n"
+                )
+            elif type(edge_type.type) is DirectRelationReference:
+                nr_of_edge_relations += 1
+                edge_relations_str.append(
+                    f"{p} -- {edge_type.direction} --> ({edge_type.source.space}, {edge_type.source.external_id}, {edge_type.source.version})\n"
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:128]:
```suggestion
                    f"{p} -- {edge_type.direction} --> ({edge_type.source.space}, {edge_type.source.external_id}, {edge_type.source.version})"
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+        nr_of_direct_relations = 0
+        nr_of_edge_relations = 0
+        for p, edge_type in view.properties.items():
+            if type(edge_type.type) is DirectRelation:
+                nr_of_direct_relations += 1
+                if edge_type.source is None:
+                    direct_relations_str.append(f"{p} --> no source")
+                    continue
+                direct_relations_str.append(
+                    f"{p} --> ({edge_type.source.space}, {edge_type.source.external_id}, {edge_type.source.version})\n"
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:123]:
```suggestion
                    f"{p} --> ({edge_type.source.space}, {edge_type.source.external_id}, {edge_type.source.version})"
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+            implements_str[-1] = implements_str[-1][0:-1]
+        table.add_row("Implements", "".join(implements_str))
+        properties = [f"{p}\n" for p in view.properties.keys()]
+        if len(properties) > 0:
+            properties[-1] = properties[-1][0:-1]
+        table.add_row("List of properties", "".join(properties))
+        direct_relations_str = []
+        edge_relations_str = []
+        nr_of_direct_relations = 0
+        nr_of_edge_relations = 0
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:115]: Lets take these one at the bottom
```suggestion
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+        for p, edge_type in view.properties.items():
+            if type(edge_type.type) is DirectRelation:
+                nr_of_direct_relations += 1
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:118]:
```suggestion
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+                if edge_type.source is None:
+                    direct_relations_str.append(f"{p} --> no source")
+                    continue
+                direct_relations_str.append(
+                    f"{p} --> ({edge_type.source.space}, {edge_type.source.external_id}, {edge_type.source.version})\n"
+                )
+            elif type(edge_type.type) is DirectRelationReference:
+                nr_of_edge_relations += 1
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:126]:
```suggestion
```

File: cognite_toolkit/cdf_tk/describe.py [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+            print(e)
+        table.add_row("Number of nodes", f"{node_count:,}")
+        edge_count = 0
+        # Iterate over all the edges in the view 1,000 at the time
+        try:
+            result = client.data_modeling.instances.aggregate(
+                view.as_id(),
+                aggregates=Count("externalId"),
+                instance_type="edge",
+            )
```
Comment (doctrino) [inline cognite_toolkit/cdf_tk/describe.py:156]: This is a bug, edges are not located in views (unless they have properties).

**Review by doctrino — COMMENTED** [review]
Many small things.

**Review by gregertw — COMMENTED** [review]
> What is the motivation for introducing another app?  

That is the pattern to introduce a new subcommand describe. I add describe datamodel, but later we can add other describe commands, like describe groups.

File: cognite_toolkit/cdf_tk/describe.py (hunk 23 above)
Comment (doctrino) [reply cognite_toolkit/cdf_tk/describe.py:142]: I realized I do not know how to do this so I have asked in the topic-data-modeling channel.

**Pull request merged**